lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs

```diff
@@ -447,10 +465,204 @@ fileModeSpec =  do
                 testTxs
                 mempty
 
+        describe "Rollback scenarios" $ do
+            let blockHash x = Hash $ x <> BS.pack (replicate (32 - (BS.length x)) 0)
+            let dummyAddr x = Address $ x <> BS.pack (replicate (32 - (BS.length x)) 0)
+            it "(NEEDS FIX, #1567) Pending txs can duplicate balance" $ \f -> do
+                (_ctx, db@DBLayer{..}) <- newDBLayer' (Just f)
+
+                let ourAddrs = knownAddresses (getState testCp)
+
+                atomically $ do
+                    unsafeRunExceptT $ initializeWallet testPk testCp testMetadata mempty
+
+                let mockApply h mockTxs = do
+                        Just cpA <- atomically $ readCheckpoint testPk
+                        let slotA = slotId $ currentTip cpA
+                        let hashA = headerHash $ currentTip cpA
+                        let fakeBlock = Block
+                                (BlockHeader
+                                { slotId = slotA { slotNumber = (slotNumber slotA) + 1}
+                                -- Assume the slotnumbers are very low
+                                , blockHeight = Quantity 0
+                                , headerHash = h
+                                , parentHeaderHash = hashA
+                                })
+                                mockTxs
+                                []
+                        let (FilteredBlock _ txs, cpB) = applyBlock fakeBlock cpA
+                        atomically $ do
+                            unsafeRunExceptT $ putCheckpoint testPk cpB
+                            unsafeRunExceptT $ putTxHistory testPk txs
+
+                let mockApplyBlock1 = mockApply (blockHash "block1")
+                            [ Tx (blockHash "tx1")
+                                [(TxIn (blockHash "faucet") 0, Coin 5)]
+                                [ TxOut (head ourAddrs) (Coin 5)
+                                ]
+                            ]
+                mockApplyBlock1
+                getAvailableBalance db `shouldReturn` 5
+
+                mockApply (blockHash "block2a")
+                            [ Tx
+                                (blockHash "tx2a")
+                                [ (TxIn (blockHash "tx1") 0, Coin 5)
+                                ]
+                                [ TxOut (dummyAddr "faucetAddr2") (Coin 3)
+                                , TxOut (ourAddrs !! 1) (Coin 2)
+                                ]
+                            ]
+
+                getAvailableBalance db `shouldReturn` 2
+                getTotalBalance db `shouldReturn` 2
+                getTxsInLedger db `shouldReturn` [(Outgoing, 3), (Incoming, 5)]
+                getPending db `shouldReturn` []
+
+
+
+                atomically $ void $ unsafeRunExceptT $ rollbackTo testPk (SlotId 0 0)
+                getTxsInLedger db `shouldReturn` []
+                -- getPending `shouldReturn` []
```

Comment:
I disagree here. getPending should return the outgoing transaction made at slot 2